core/scenegraph/src/main/java/edu/cmu/cs/dennisc/scenegraph/Component.java

```diff
@@ -238,37 +237,107 @@ protected void fireHierarchyChanged() {
     fireHierarchyChanged(new HierarchyEvent(this));
   }
 
+  // Vector4
+  public Vector4 transformToAbsolute(Vector4 xyz) {
+    Vector4 rv = new Vector4(xyz);
+    if (!isAbsolute()) {
+      AffineMatrix4x4 m = getAbsoluteTransformation();
+      m.transform(rv);
+    }
+    return rv;
+  }
+
+  public Vector4 transformFromAbsolute(Vector4 xyz) {
+    Vector4 rv = new Vector4(xyz);
+    if (!isAbsolute()) {
+      AffineMatrix4x4 m = getInverseAbsoluteTransformation();
+      m.transform(rv);
+    }
+    return rv;
+  }
+
   public Vector4 transformTo(Vector4 xyzw, Component to) {
-    return TransformationUtilities.transformTo(xyzw, this, to);
+    Vector4 p = transformToAbsolute(xyzw);
+    return to.transformFromAbsolute(p);
   }
 
   public Vector4 transformFrom(Vector4 xyzw, Component from) {
     return from.transformTo(xyzw, this);
   }
 
+  // Vector3
+  public Vector3 transformToAbsolute(Vector3 xyz) {
+    Vector3 rv = new Vector3(xyz);
+    if (!isAbsolute()) {
+      AffineMatrix4x4 m = getAbsoluteTransformation();
+      m.transform(rv);
+    }
+    return rv;
+  }
+
+  public Vector3 transformFromAbsolute(Vector3 xyz) {
+    Vector3 rv = new Vector3(xyz);
+    if (!isAbsolute()) {
+      AffineMatrix4x4 m = getInverseAbsoluteTransformation();
+      m.transform(rv);
+    }
+    return rv;
+  }
+
+  public Vector3 transformTo(Vector3 xyz, Component to) {
+    Vector3 p = transformToAbsolute(xyz);
+    return to.transformFromAbsolute(p);
+  }
+
+  //Point3d
+  public Point3 transformToAbsolute(Point3 xyz) {
+    Point3 rv = new Point3(xyz);
+    if (!isAbsolute()) {
+      AffineMatrix4x4 m = getAbsoluteTransformation();
+      m.transform(rv);
+    }
+    return rv;
+  }
+
+  public Point3 transformFromAbsolute(Point3 xyz) {
+    Point3 rv = new Point3(xyz);
+    if (!isAbsolute()) {
+      AffineMatrix4x4 m = getInverseAbsoluteTransformation();
+      m.transform(rv);
+    }
+    return rv;
+  }
+
   public Point3 transformTo(Point3 xyz, Component to) {
-    return TransformationUtilities.transformTo(xyz, this, to);
+    Point3 p = transformToAbsolute(xyz);
+    return to.transformFromAbsolute(p);
   }
 
   public Point3 transformFrom(Point3 xyz, Component from) {
     return from.transformTo(xyz, this);
   }
 
-  public Vector3 transformTo(Vector3 xyz, Component to) {
-    return TransformationUtilities.transformTo(xyz, this, to);
-  }
 
   public Vector3 transformFrom(Vector3 xyz, Component from) {
     return from.transformTo(xyz, this);
   }
 
+  private boolean isAbsolute() {
+    //todo: call isSceneOf()?
+    if ((this instanceof Scene) || ((ReferenceFrame) this == AsSeenBy.SCENE)) {
```

Comment:
Now that this is on Component the ReferenceFrame is more obviously not the case, so it simplifies to `instanceOf`, or just false with an overriding true in Scene.